subsys/logging/log_backend_adsp_mtrace.c

```diff
@@ -0,0 +1,252 @@
+/*
+ * Copyright (c) 2019,2022 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/logging/log_backend_adsp_mtrace.h>
+#include <zephyr/logging/log_backend.h>
+#include <zephyr/logging/log_core.h>
+#include <zephyr/logging/log_output.h>
+#include <zephyr/logging/log_backend_std.h>
+#include <zephyr/zephyr.h>
+#include <soc.h>
+#include <adsp_memory.h>
+
+/*
+ * A lock is needed as log_process() and log_panic() have no internal locks
+ * to prevent concurrency. Meaning if log_process is called after
+ * log_panic was called previously, log_process may happen from another
+ * CPU and calling context than the log processing thread running in
+ * the background. On an SMP system this is a race.
+ *
+ * This caused a race on the output trace such that the logging output
+ * was garbled and useless.
+ */
+static struct k_spinlock mtrace_lock;
+
+static uint32_t log_format_current = CONFIG_LOG_BACKEND_ADSP_MTRACE_OUTPUT_DEFAULT;
+
+static adsp_mtrace_log_hook_t mtrace_hook;
+
+static bool mtrace_active;
+
+/*
+ * SRAM window for mtrace is organized in 16 (equal sized) pages:
+ *
+ *	------------------------
+ *	| Page0  - descriptors |
+ *	------------------------
+ *	| Page1  - slot0       |
+ *	------------------------
+ *	| Page2  - slot1       |
+ *	------------------------
+ *	|         ...          |
+ *	------------------------
+ *	| Page14  - slot13     |
+ *	------------------------
+ *	| Page15  - slot14     |
+ *	------------------------
+ *
+ * The slot size == page size
+ *
+ * The first page contains descriptors for the remaining 15 cores
+ * The slot descriptor is:
+ * u32 res_id;
+ * u32 type;
+ * u32 vma;
+ *
+ * Log buffer slots have the following layout:
+ * u32 host_read_ptr;
+ * u32 dsp_write_ptr;
```

Comment:
Would be good to see a state diagram or something for how this protocol gets used.  It looks to me like those ptrs are implemented here to mean "the next byte that {host,dsp} will {read,write}"?  But if that's the case, then how do you disambiguate "empty" vs. "full" when the two pointers are equal?  At startup, they are both zero ("empty").  But later on, if the dsp has filled the buffer, they might be equal again ("full").

Reply:
Also a design note, though that would require changes to host firmware: if you put those pointers into the descriptor space you can both reclaim otherwise-wasted words and also implement the slot data as a straightforward power-of-two ring buffer.

Reply:
Thanks, protocol change is not possible, I'd rather just use ADSP backend in that case.

Let me add some details on the protocol in next update. Buffer is empty if "read==write" (i.e. starts empty). It is full when "read==write+1" (module buffer size). I'm testing this primarily against https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sof/ipc4-mtrace.c and it follows this convention.

This is not as nice as winstream, but again we need to stick to the protocol.